src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -0,0 +1,415 @@
+        /// <param name="isSpecialToken">Indicate if the token is a special token.</param>
+        /// <param name="accumulatedIds">The list of accumulated Ids.</param>
+        /// <returns>True if the operation succeeded, false otherwise.</returns>
+        public override bool TokenizeToIds(string sequence, bool isSpecialToken, List<int> accumulatedIds)
+        {
+            if (string.IsNullOrEmpty(sequence))
+            {
+                return true;
+            }
+
+            if (isSpecialToken)
+            {
+                if (_specialTokensEncoder is not null && _specialTokensEncoder.TryGetValue(sequence, out int id))
+                {
+                    accumulatedIds.Add(id);
+                    return true;
+                }
+
+                return false;
+            }
+
+            if (_cache.Lookup(sequence, out int[] tokenIds))
+            {
+                accumulatedIds.AddRange(tokenIds);
+                return true;
+            }
+
+            if (_vocab.TryGetValue(sequence, out int mappedId))
+            {
+                accumulatedIds.Add(mappedId);
+                return true;
+            }
+
+            int[] encodedIds = BytePairEncoder.BytePairEncode(Encoding.UTF8.GetBytes(sequence), _encoder);
+            _cache.Add(sequence, encodedIds);
+
+            accumulatedIds.AddRange(encodedIds);
+            return true;
+        }
+
+        /// <summary>
+        /// Map the token to tokenized Id.
+        /// </summary>
+        /// <param name="token">The token to map to the Id.</param>
+        /// <returns>The mapped Id of the token.</returns>
+        public override int? TokenToId(string token) => TokenToId(token, skipSpecialTokens: false);
+
+        /// <summary>
+        /// Map the token to tokenized Id.
+        /// </summary>
+        /// <param name="token">The token to map to the Id.</param>
+        /// <param name="skipSpecialTokens">Indicate if want to skip the special tokens during the encoding.</param>
+        /// <returns>The mapped Id of the token.</returns>
+        public override int? TokenToId(string token, bool skipSpecialTokens)
+        {
+            if (string.IsNullOrEmpty(token))
+            {
+                return 0;
+            }
+
+            if (!skipSpecialTokens && _specialTokensEncoder is not null && _specialTokensEncoder.TryGetValue(token, out int specialTokenId))
+            {
+                return specialTokenId;
+            }
+
+            if (_cache.Lookup(token, out int[] ids))
+            {
+                if (ids.Length == 1)
+                {
+                    return ids[0];
+                }
+
+                return null;
+            }
+
+            if (_vocab.TryGetValue(token, out int id))
+            {
+                return id;
+            }
+
+            int[] idsToCache = BytePairEncoder.BytePairEncode(Encoding.UTF8.GetBytes(token), _encoder);
+            _cache.Add(token, idsToCache.ToArray());
```

Comment:
Why does this need to create a new array?

Reply:
This is left over from the code I picked from MS Library. But it is not needed after my changes I have done. Thanks for the catch.